File: apps/desktop/src/native/api/rack.ts
import type { ArrangementMutationResult, BuiltInInstrumentSummary } from '@/model/domain';
import { invokeHost } from '../invoke';

async function invokeRack(command: string, args: Record<string, unknown>) {
  return invokeHost<ArrangementMutationResult>(command, args);
}

export async function listBuiltInInstruments(): Promise<BuiltInInstrumentSummary[]> {
  return await invokeHost<BuiltInInstrumentSummary[]>('list_built_in_instruments', {});
}

export async function setTrackBuiltInInstrument(
  trackId: string,
  presetId: string,
): Promise<ArrangementMutationResult> {
  return await invokeRack('set_track_built_in_instrument', { trackId, presetId });
}

export async function setTrackVst3Instrument(
  trackId: string,
  pluginPath: string,
): Promise<ArrangementMutationResult> {
  return await invokeRack('set_track_vst3_instrument', { trackId, pluginPath });
}

export async function clearTrackInstrument(trackId: string): Promise<ArrangementMutationResult> {
  return await invokeRack('clear_track_instrument', { trackId });
}

export async function addTrackEffect(
  trackId: string,
  pluginPath: string,
): Promise<ArrangementMutationResult> {
  return await invokeRack('add_track_effect', { trackId, pluginPath });
}

export async function removeTrackEffect(
  trackId: string,
  deviceId: string,
): Promise<ArrangementMutationResult> {
  return await invokeRack('remove_track_effect', { trackId, deviceId });
}

export async function reorderTrackEffects(
  trackId: string,
  orderedDeviceIds: string[],
): Promise<ArrangementMutationResult> {
  return await invokeRack('reorder_track_effects', { trackId, orderedDeviceIds });
}

export async function setTrackDeviceBypassed(
  trackId: string,
  deviceId: string,
  bypassed: boolean,
): Promise<ArrangementMutationResult> {
  return await invokeRack('set_track_device_bypassed', { trackId, deviceId, bypassed });
}

export async function setTrackDeviceParameter(
  trackId: string,
  deviceId: string,
  parameterIndex: number,
  value: number,
): Promise<ArrangementMutationResult> {
  return await invokeRack('set_track_device_parameter', {
    trackId,
    deviceId,
    parameterIndex,
    value,
  });
}

export async function openTrackPluginEditor(trackId: string, deviceId: string): Promise<void> {
  await invokeHost<void>('open_track_plugin_editor', { trackId, deviceId });
}

File: apps/desktop/src-tauri/src/session/commands/rack.rs
use super::*;

#[tauri::command]
pub async fn set_track_instrument(
    track_id: String,
    plugin_path: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "instrument.set",
        json!({ "trackId": track_id, "pluginPath": plugin_path }),
    )
    .await
}

#[tauri::command]
pub async fn clear_track_instrument(
    track_id: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(app, "instrument.clear", json!({ "trackId": track_id })).await
}

#[tauri::command]
pub async fn add_track_effect(
    track_id: String,
    plugin_path: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "effect.add",
        json!({ "trackId": track_id, "pluginPath": plugin_path }),
    )
    .await
}

#[tauri::command]
pub async fn remove_track_effect(
    track_id: String,
    device_id: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "effect.remove",
        json!({ "trackId": track_id, "deviceId": device_id }),
    )
    .await
}

#[tauri::command]
pub async fn reorder_track_effects(
    track_id: String,
    ordered_device_ids: Vec<String>,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "effect.reorder",
        json!({ "trackId": track_id, "deviceIds": ordered_device_ids }),
    )
    .await
}

#[tauri::command]
pub async fn set_track_device_bypassed(
    track_id: String,
    device_id: String,
    bypassed: bool,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "device.bypass",
        json!({ "trackId": track_id, "deviceId": device_id, "bypassed": bypassed }),
    )
    .await
}

#[tauri::command]
pub async fn set_track_device_parameter(
    track_id: String,
    device_id: String,
    parameter_index: u32,
    value: f32,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "device.parameter.set",
        json!({
            "trackId": track_id,
            "deviceId": device_id,
            "parameterIndex": parameter_index,
            "value": value,
        }),
    )
    .await
}

#[tauri::command]
pub async fn open_track_plugin_editor(
    track_id: String,
    device_id: String,
    app: AppHandle,
) -> Result<(), String> {
    dispatch(
        app,
        "plugin.editor.open",
        json!({ "trackId": track_id, "deviceId": device_id }),
    )
    .await
}
